File: ios-samples/poslink/EposPosLinkExample/EposPosLinkExample/Transactions/TransactionRecorder.swift
import Foundation
import TeyaUnifiedEposSDK

enum TransactionRecorder {

    static func recordPaymentIfFinal(
        state: TeyaPaymentStateDetails,
        type: TeyaTransactionType,
        isTab: Bool = false,
        isCash: Bool = false
    ) {
        guard state.isFinal else { return }
        TransactionStore.shared.upsert(
            TransactionRecord(
                id: state.eposTransactionId,
                type: type,
                isSuccess: state.state == .successful,
                amountMinor: Int(state.amount),
                currency: state.currency,
                gatewayPaymentId: state.gatewayPaymentId?.id,
                timestamp: state.transactionTimestamp?.asInt64 ?? Int64(Date().timeIntervalSince1970 * 1000),
                isCashPayment: isCash,
                isTabPayment: isTab
            )
        )
    }

    static func recordRefund(
        refundResult: TeyaRefundResultDetails,
        gatewayPaymentId: String,
        amountMinor: Int32,
        currency: String
    ) {
        TransactionStore.shared.upsert(
            TransactionRecord(
                id: refundResult.gatewayRefundId?.id ?? UUID().uuidString,
                type: .refund,
                isSuccess: refundResult.result == TeyaRefundResult.success,
                amountMinor: Int(amountMinor),
                currency: currency,
                gatewayPaymentId: nil,
                timestamp: Int64(Date().timeIntervalSince1970 * 1000)
            )
        )
        if refundResult.result == TeyaRefundResult.success {
            TransactionStore.shared.markRefunded(gatewayPaymentId: gatewayPaymentId)
        }
    }
}

File: windows-samples/poslink/EposPosLinkExample/EposPosLinkExample/Helpers/TransactionRecorder.cs
using System;
using EposPosLinkExample.Models;

namespace EposPosLinkExample.Helpers
{
    public class TransactionRecorder
    {
        private readonly TransactionStore _transactionStore = TransactionStore.Instance;

        public void RecordPaymentIfFinal(MakePaymentStateChange state, TransactionType type, bool isTab = false, bool isCash = false)
        {
            if (!state.IsFinal) return;

            _transactionStore.Upsert(new TransactionRecord(
                Id: state.TransactionId,
                Type: type,
                IsSuccess: state.State == MakePaymentStateChange.PaymentState.SUCCESSFUL,
                AmountMinor: state.Amount,
                Currency: state.Currency,
                GatewayPaymentId: state.GatewayPaymentId,
                Timestamp: state.TransactionTimestamp ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds())
            {
                IsCashPayment = isCash,
                IsTabPayment = isTab
            });
        }

        public void RecordRefund(RefundResult result, string gatewayPaymentId, int refundAmount, string currency)
        {
            _transactionStore.Upsert(new TransactionRecord(
                Id: result.GatewayRefundId ?? Guid.NewGuid().ToString(),
                Type: TransactionType.Refund,
                IsSuccess: result.IsSuccess,
                AmountMinor: refundAmount,
                Currency: currency,
                GatewayPaymentId: null,
                Timestamp: DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()));

            if (result.IsSuccess)
            {
                _transactionStore.MarkRefunded(gatewayPaymentId);
            }
        }
    }
}